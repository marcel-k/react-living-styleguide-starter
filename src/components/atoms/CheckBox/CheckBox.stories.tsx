import { action } from '@storybook/addon-actions';
import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { Dispatch, SetStateAction } from 'react';
import { useState } from 'react';

import CheckBox from './CheckBox';

const handleAction = (setChecked: Dispatch<SetStateAction<boolean>>) =>
  (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => {
    setChecked(checked);
    return action(`Change to ${checked}`)(event);
  };

storiesOf('atoms/CheckBox', module)
  .add('default', () => {
    const [checked, setChecked] = useState(false);
    return (
      <CheckBox
        checked={checked}
        onChange={handleAction(setChecked)}
      />
    );
  })
  .add('default checked', () => {
    const [checked, setChecked] = useState(true);
    return (
      <CheckBox
        checked={checked}
        onChange={handleAction(setChecked)}
      />
    );
  })
  .add('with label', () => {
    const [checked, setChecked] = useState(false);
    return (
      <CheckBox
        checked={checked}
        onChange={handleAction(setChecked)}
        label={text('label', 'this is a label')}
      />
    );
  })
  .add('disabled', () => {
    const [checked, setChecked] = useState(false);
    const [disabled] = useState(true);
    return (
      <CheckBox
        checked={checked}
        label={`can't touch this`}
        onChange={handleAction(setChecked)}
        disabled={boolean('disabled', disabled)}
      />
    );
  });