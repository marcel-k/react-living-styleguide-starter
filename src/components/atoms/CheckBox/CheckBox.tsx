import React from 'react';
import { useRef } from 'react';
import { ChangeEvent } from 'react';
import styled from 'styled-components/macro';

interface CheckBoxDeclProps {
  /**
   * Sets the state of the checkBox.
   * @default false
   */
  checked?: boolean;
  /**
  * Fires when the checkbox checked state wants to change.
  */
  onChange?: (event: ChangeEvent<HTMLInputElement>, checked: boolean) => void;
}

interface LabelProps {
  /**
  * Give the checkbox a (clickable) label.
  */
 label?: string;
}

export interface CheckBoxProps extends CheckBoxDeclProps, LabelProps {
  /**
   * Set if the checkbox is disabled or not.
   * @default false
   */
  disabled?: boolean;
}

/**
 * Create an invisible checkbox that is visually overriden.
 */
const CheckBoxDecl = styled.input.attrs((props: CheckBoxDeclProps) => ({
  type: 'checkbox',
  checked: props.checked,
  onChange: props.onChange
}))`
  /* empty */
  `;

const StyledCheckBox = styled(CheckBoxDecl)`
    top: 0;
    left: 0;
    width: 100%;
    cursor: inherit;
    height: 100%;
    margin: 0;
    opacity: 0;
    padding: 0;
    position: absolute;
`;

const ClickArea = styled.span`
    display: inline-flex;
    justify-content:center;
    align-items: center;
    width: 40px; /* TODO: theme */
    height: 40px; /* TODO: theme */
    margin: 0;
    cursor: pointer;
    position: relative;
    user-select: none;
`;

const ClickBox = styled.span`
  display: inline-flex;
  justify-content:center;
  align-items: center;
  width: 28px; /* TODO: theme - border? */
  height: 28px; /* TODO: theme - border? */
  color: #bbb; /* TODO: theme */
  text-align: center;
  border: 2px solid #bbb; /* TODO: them */
  cursor: pointer;
  user-select: none;
  text-decoration: none;
`;

const Wrapper = styled.span<{ disabled: boolean }>`
    cursor: pointer;
    display: inline-flex;
    align-items: center;
    vertical-align: top;
    opacity: ${(props) => props.disabled ? 0.5 : 1}; /* TODO: theme */
`;

const Label = styled.label`
    display: block;
    margin: 0;
    color: #333; /* TODO: theme */
    font-size: 0.875rem; /* TODO: theme */
    font-weight: 700; /* TODO: theme */
    cursor: pointer;
`;

/**
 * A simple Checkbox component
 */
const CheckBox = (props: CheckBoxProps) => {

  const {
    label = null,
    checked = false,
    disabled = false,
    onChange = () => { },
    ...inputProps
  } = props;

  const input = useRef<HTMLInputElement>(null);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    if (!disabled) {
      onChange(event, event.target.checked);
    }
  };

  const handleLabelClick = () => {
    if (input && input.current) {
      input.current.click();
    }
  }

  return (
    <Wrapper disabled={disabled}>
      <ClickArea>
        <StyledCheckBox
          ref={input}
          checked={checked}
          onChange={handleChange}
          {...inputProps}
        />
        <ClickBox>
          {/* TODO: theme 0000ff */}
          {checked &&
            <img src={`https://img.icons8.com/metro/26/${'0000ff'}/checkmark.png`} />
          }
        </ClickBox>
      </ClickArea>
      {label &&
        <Label onClick={handleLabelClick}>
          {label}
        </Label>
      }
    </Wrapper>
  );
};

// export default forwardRef(CheckBox);
export default CheckBox;