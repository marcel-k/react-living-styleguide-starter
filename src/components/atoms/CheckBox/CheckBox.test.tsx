import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import 'jest-styled-components';
import 'react-testing-library/cleanup-after-each';

import CheckBox from './CheckBox';

describe('<CheckBox /> spec', () => {
  it('renders default unchecked', function () {
    const component = (<CheckBox />);
    const { container } = render(component);
    const { firstElementChild } = container;

    const hiddenInput = (firstElementChild as Element).querySelector('input') as HTMLInputElement;

    expect(hiddenInput).toHaveProperty('checked', false);
  });

  it('renders with label', function () {
    const component = (<CheckBox label={'this is a label'} />);
    const { container } = render(component);
    const { firstElementChild } = container;

    const label = (firstElementChild as Element).querySelector('label') as HTMLLabelElement;

    expect(label.textContent).toEqual('this is a label');
  });

  it('shows a checkmark when checked', function () {
    const spy = jest.fn();
    const component = (
      <CheckBox
        onChange={spy}
        label={'this is a label'}
      />
    );
    const { container, rerender } = render(component);

    const { firstElementChild } = container;
    expect(firstElementChild).toMatchSnapshot(); // unchecked

    const hiddenInput = (firstElementChild as Element).querySelector('input') as HTMLInputElement;
    fireEvent.click(hiddenInput as Element);
    expect(spy).toHaveBeenCalledTimes(1);

    rerender(
      <CheckBox
        onChange={spy}
        checked={true}
        label={'this is a label'}
      />);

    expect(firstElementChild).toMatchSnapshot(); // checked
  });
});