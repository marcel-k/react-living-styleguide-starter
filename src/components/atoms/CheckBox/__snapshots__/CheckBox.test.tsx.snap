// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckBox /> spec shows a checkmark when checked 1`] = `
.c2 {
  top: 0;
  left: 0;
  width: 100%;
  cursor: inherit;
  height: 100%;
  margin: 0;
  opacity: 0;
  padding: 0;
  position: absolute;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 40px;
  height: 40px;
  margin: 0;
  cursor: pointer;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 28px;
  height: 28px;
  color: #bbb;
  text-align: center;
  border: 2px solid #bbb;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 {
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: top;
  opacity: 1;
}

.c4 {
  display: block;
  margin: 0;
  color: #333;
  font-size: 0.875rem;
  font-weight: 700;
  cursor: pointer;
}

<span
  class="c0"
>
  <span
    class="c1"
  >
    <input
      class="c2"
      type="checkbox"
    />
    <span
      class="c3"
    />
  </span>
  <label
    class="c4"
  >
    this is a label
  </label>
</span>
`;

exports[`<CheckBox /> spec shows a checkmark when checked 2`] = `
<span
  class="CheckBox__Wrapper-iggjvv-4 feVCkj"
>
  <span
    class="CheckBox__ClickArea-iggjvv-2 djdmne"
  >
    <input
      class="CheckBox__CheckBoxDecl-iggjvv-0 CheckBox__StyledCheckBox-iggjvv-1 blKBhu"
      type="checkbox"
    />
    <span
      class="CheckBox__ClickBox-iggjvv-3 dqJzvL"
    >
      <img
        src="https://img.icons8.com/metro/26/0000ff/checkmark.png"
      />
    </span>
  </span>
  <label
    class="CheckBox__Label-iggjvv-5 eurPZR"
  >
    this is a label
  </label>
</span>
`;
